src/Core/BufferLine.re

```diff
@@ -33,24 +37,130 @@ type t = {
   indentation: IndentationSettings.t,
   // [raw] is the raw string (byte array)
   raw: string,
+  // [font] is the main font to render the line with
+  font: Font.t,
+  // [skiaTypeface] is the Skia typeface associated with the font
+  skiaTypeface: Skia.Typeface.t,
+  // [glyphStrings] is a list of typefaces and strings from shaping
+  mutable glyphStrings: list((Skia.Typeface.t, string)),
   // [characters] is a cache of discovered characters we've found in the string so far
   mutable characters: array(option(characterCacheInfo)),
   // [byteIndexMap] is a cache of byte -> index
   mutable byteIndexMap: array(option(int)),
-  // nextByte is the nextByte to work from, or -1 if complete
+  // [nextByte] is the nextByte to work from, or -1 if complete
   mutable nextByte: int,
-  // nextIndex is the nextIndex to work from
+  // [nextIndex] is the nextIndex to work from
   mutable nextIndex: int,
-  // nextPosition is the graphical position (based on character width)
-  mutable nextPosition: int,
+  // [nextGlyphStringIndex] is the next nth glyph string in shapes
+  mutable nextGlyphStringByte: int,
+  // [nextCharacterPosition] is the graphical position (based on character width)
+  mutable nextCharacterPosition: int,
+  // [nextPixelPosition] is the graphical position (based on pixels)
+  mutable nextPixelPosition: float,
 };
 
 module Internal = {
-  let measure = (indentationSettings: IndentationSettings.t, c) =>
-    if (Uchar.equal(c, tab)) {
-      indentationSettings.tabSize;
+  let skiaTypefaceEqual = (tf1, tf2) =>
+    Skia.Typeface.getUniqueID(tf1) == Skia.Typeface.getUniqueID(tf2);
+
+  // We want to cache measurements tied to typefaces and characters, since text measuring is expensive
+  module SkiaTypefaceUcharHashable = {
+    // This is of the form typeface, size, uchar
+    type t = (Skia.Typeface.t, float, Uchar.t);
+
+    let equal = ((tf1, s1, uc1), (tf2, s2, uc2)) =>
+      skiaTypefaceEqual(tf1, tf2) && s1 == s2 && Uchar.equal(uc1, uc2);
+
+    let hash = ((tf, s, uc)) =>
+      Int32.to_int(Skia.Typeface.getUniqueID(tf))
+      + Uchar.hash(uc)
+      + Hashtbl.hash(s);
+  };
+
+  module MeasureResult = {
+    type t = float;
+
+    let weight = _ => 1;
+  };
+
+  module MeasurementsCache =
+    Lru.M.Make(SkiaTypefaceUcharHashable, MeasureResult);
+
+  let measurementsCache =
+    MeasurementsCache.create(~initialSize=1024, 128 * 1024);
+
+  // Create a paint to measure the character with
+  let paint = Skia.Paint.make();
+  Skia.Paint.setTextEncoding(paint, GlyphId);
+  Skia.Paint.setLcdRenderText(paint, true);
+  Skia.Paint.setAntiAlias(paint, true);
```

Comment:
We'll need to make sure these stay in sync with the user settings - in particular, the `setSubpixelText` can impact rendering and cause sizes to be off - and can be changed with `editor.fontSmoothing` configuration. 

This inconsistency caused problems like #1364 (fixed in https://github.com/onivim/oni2/pull/1372)

Reply:
Ah thanks! I set the smoothing every time I measure, but it looks like I should get rid of that `antiAlias` call:

https://github.com/revery-ui/revery/blob/1531d5feebb894830010b14dadf89e09db9b648b/src/Font/Smoothing.re#L12

Will do now!

Reply:
Fixed in 857b541e05b1018f561985ad8917e3fa546a8cdf!